File: TestVectorsAwsCryptographicMaterialProviders/runtimes/python/__main__-py/StormTracker.py
import sys
from typing import Callable, Any, TypeVar, NamedTuple
from math import floor
from itertools import count

import module_
import _dafny
import System_
import Wrappers
import Relations
import Seq_MergeSort
import Math
import Seq
import BoundedInts
import Unicode
import Functions
import Utf8EncodingForm
import Utf16EncodingForm
import UnicodeStrings
import FileIO
import GeneralInternals
import MulInternalsNonlinear
import MulInternals
import Mul
import ModInternalsNonlinear
import DivInternalsNonlinear
import ModInternals
import DivInternals
import DivMod
import Power
import Logarithm
import StandardLibrary_UInt
import StandardLibrary_String
import StandardLibrary
import UUID
import UTF8
import Time
import Streams
import Sorting
import SortedSets
import HexStrings
import FloatCompare
import ConcurrentCall
import Base64
import Base64Lemmas
import Actions
import DafnyLibraries
import software_amazon_cryptography_services_dynamodb_internaldafny_types
import software_amazon_cryptography_services_kms_internaldafny_types
import software_amazon_cryptography_keystore_internaldafny_types
import software_amazon_cryptography_primitives_internaldafny_types
import software_amazon_cryptography_materialproviders_internaldafny_types
import AlgorithmSuites
import Materials
import Keyring
import MultiKeyring
import AwsArnParsing
import AwsKmsMrkAreUnique
import AwsKmsMrkMatchForDecrypt
import AwsKmsUtils
import Constants
import ExternRandom
import Random
import AESEncryption
import ExternDigest
import Digest
import HMAC
import WrappedHMAC
import HKDF
import WrappedHKDF
import Signature
import KdfCtr
import RSAEncryption
import AwsCryptographyPrimitivesOperations
import software_amazon_cryptography_primitives_internaldafny
import Aws_Cryptography
import Aws
import MaterialWrapping
import CanonicalEncryptionContext
import IntermediateKeyWrapping
import EdkWrapping
import AwsKmsKeyring
import StrictMultiKeyring
import AwsKmsDiscoveryKeyring
import software_amazon_cryptography_services_kms_internaldafny
import DiscoveryMultiKeyring
import AwsKmsMrkDiscoveryKeyring
import MrkAwareDiscoveryMultiKeyring
import AwsKmsMrkKeyring
import MrkAwareStrictMultiKeyring
import LocalCMC
import software_amazon_cryptography_internaldafny_SynchronizedLocalCMC

# Module: StormTracker

class default__:
    def  __init__(self):
        pass

    @staticmethod
    def DefaultStorm():
        return software_amazon_cryptography_materialproviders_internaldafny_types.StormTrackingCache_StormTrackingCache(1000, Wrappers.Option_Some(1), 10, 1, 20, 20, 20)


class CacheState:
    @classmethod
    def default(cls, ):
        return lambda: CacheState_EmptyWait()
    def __ne__(self, __o: object) -> bool:
        return not self.__eq__(__o)
    @property
    def is_EmptyWait(self) -> bool:
        return isinstance(self, CacheState_EmptyWait)
    @property
    def is_EmptyFetch(self) -> bool:
        return isinstance(self, CacheState_EmptyFetch)
    @property
    def is_Full(self) -> bool:
        return isinstance(self, CacheState_Full)

class CacheState_EmptyWait(CacheState, NamedTuple('EmptyWait', [])):
    def __dafnystr__(self) -> str:
        return f'StormTracker.CacheState.EmptyWait'
    def __eq__(self, __o: object) -> bool:
        return isinstance(__o, CacheState_EmptyWait)
    def __hash__(self) -> int:
        return super().__hash__()

class CacheState_EmptyFetch(CacheState, NamedTuple('EmptyFetch', [])):
    def __dafnystr__(self) -> str:
        return f'StormTracker.CacheState.EmptyFetch'
    def __eq__(self, __o: object) -> bool:
        return isinstance(__o, CacheState_EmptyFetch)
    def __hash__(self) -> int:
        return super().__hash__()

class CacheState_Full(CacheState, NamedTuple('Full', [('data', Any)])):
    def __dafnystr__(self) -> str:
        return f'StormTracker.CacheState.Full({_dafny.string_of(self.data)})'
    def __eq__(self, __o: object) -> bool:
        return isinstance(__o, CacheState_Full) and self.data == __o.data
    def __hash__(self) -> int:
        return super().__hash__()


class StormTracker:
    def  __init__(self):
        self.wrapped: LocalCMC.LocalCMC = None
        self.inFlight: DafnyLibraries.MutableMap = None
        self.gracePeriod: int = None
        self.graceInterval: int = None
        self.fanOut: int = None
        self.inFlightTTL: int = None
        self.lastPrune: int = None
        self.sleepMilli: int = None
        pass

    def __dafnystr__(self) -> str:
        return "StormTracker.StormTracker"
    def ctor__(self, cache):
        nw31_ = LocalCMC.LocalCMC()
        nw31_.ctor__((cache).entryCapacity, ((cache).entryPruningTailSize).UnwrapOr(1))
        (self).wrapped = nw31_
        nw32_ = DafnyLibraries.MutableMap()
        (self).inFlight = nw32_
        (self).gracePeriod = (cache).gracePeriod
        (self).graceInterval = (cache).graceInterval
        (self).fanOut = (cache).fanOut
        (self).inFlightTTL = (cache).inFlightTTL
        (self).sleepMilli = (cache).sleepMilli
        (self).lastPrune = 0

    def InFlightSize(self):
        d_695_x_ = (self.inFlight).Size()
        if (d_695_x_) <= (StandardLibrary_UInt.default__.INT64__MAX__LIMIT):
            return d_695_x_
        elif True:
            return StandardLibrary_UInt.default__.INT64__MAX__LIMIT

    def FanOutReached(self, now):
        res: bool = False
        (self).PruneInFlight(now)
        res = (self.fanOut) <= ((self).InFlightSize())
        return res
        return res

    def AddLong(self, x, y):
        if (x) < ((StandardLibrary_UInt.default__.INT64__MAX__LIMIT) - (y)):
            return (x) + (y)
        elif True:
            return StandardLibrary_UInt.default__.INT64__MAX__LIMIT

    def CheckInFlight(self, identifier, result, now):
        output: CacheState = CacheState.default()()
        d_696_fanOutReached_: bool
        out136_: bool
        out136_ = (self).FanOutReached(now)
        d_696_fanOutReached_ = out136_
        if d_696_fanOutReached_:
            output = CacheState_Full(result)
            return output
        elif ((result).expiryTime) <= (now):
            out137_: CacheState
            out137_ = (self).CheckNewEntry(identifier, now)
            output = out137_
        elif (now) < (((result).expiryTime) - (self.gracePeriod)):
            output = CacheState_Full(result)
            return output
        elif True:
            if (self.inFlight).HasKey(identifier):
                d_697_entry_: int
                d_697_entry_ = (self.inFlight).Select(identifier)
                if ((self).AddLong(d_697_entry_, self.graceInterval)) > (now):
                    output = CacheState_Full(result)
                    return output
            (self.inFlight).Put(identifier, now)
            output = CacheState_EmptyFetch()
            return output
        return output

    def PruneInFlight(self, now):
        if (self.fanOut) > ((self).InFlightSize()):
            return
        if (self.lastPrune) == (now):
            return
        (self).lastPrune = now
        d_698_keySet_: _dafny.Set
        d_698_keySet_ = (self.inFlight).Keys()
        d_699_keys_: _dafny.Seq
        d_699_keys_ = SortedSets.default__.SetToSequence(d_698_keySet_)
        hi8_ = len(d_699_keys_)
        for d_700_i_ in range(0, hi8_):
            d_701_v_: int
            d_701_v_ = (self.inFlight).Select((d_699_keys_)[d_700_i_])
            if (now) >= ((self).AddLong(d_701_v_, self.inFlightTTL)):
                (self.inFlight).Remove((d_699_keys_)[d_700_i_])

    def CheckNewEntry(self, identifier, now):
        output: CacheState = CacheState.default()()
        d_702_fanOutReached_: bool
        out138_: bool
        out138_ = (self).FanOutReached(now)
        d_702_fanOutReached_ = out138_
        if d_702_fanOutReached_:
            output = CacheState_EmptyWait()
            return output
        elif (self.inFlight).HasKey(identifier):
            d_703_entry_: int
            d_703_entry_ = (self.inFlight).Select(identifier)
            if ((self).AddLong(d_703_entry_, self.graceInterval)) > (now):
                output = CacheState_EmptyWait()
                return output
        (self.inFlight).Put(identifier, now)
        output = CacheState_EmptyFetch()
        return output
        return output

    def GetFromCacheWithTime(self, input, now):
        output: Wrappers.Result = Wrappers.Result.default(CacheState.default())()
        d_704_result_: Wrappers.Result
        out139_: Wrappers.Result
        out139_ = (self.wrapped).GetCacheEntryWithTime(input, now)
        d_704_result_ = out139_
        if (d_704_result_).is_Success:
            d_705_newResult_: CacheState
            out140_: CacheState
            out140_ = (self).CheckInFlight((input).identifier, (d_704_result_).value, now)
            d_705_newResult_ = out140_
            output = Wrappers.Result_Success(d_705_newResult_)
            return output
        elif ((d_704_result_).error).is_EntryDoesNotExist:
            d_706_newResult_: CacheState
            out141_: CacheState
            out141_ = (self).CheckNewEntry((input).identifier, now)
            d_706_newResult_ = out141_
            output = Wrappers.Result_Success(d_706_newResult_)
            return output
        elif True:
            output = Wrappers.Result_Failure((d_704_result_).error)
            return output
        return output

    def GetFromCache(self, input):
        output: Wrappers.Result = Wrappers.Result.default(CacheState.default())()
        d_707_now_: int
        out142_: int
        out142_ = Time.default__.CurrentRelativeTime()
        d_707_now_ = out142_
        out143_: Wrappers.Result
        out143_ = (self).GetFromCacheWithTime(input, d_707_now_)
        output = out143_
        return output

    def GetCacheEntry(self, input):
        output: Wrappers.Result = None
        d_708_result_: Wrappers.Result
        out144_: Wrappers.Result
        out144_ = (self).GetFromCache(input)
        d_708_result_ = out144_
        if (d_708_result_).is_Failure:
            output = Wrappers.Result_Failure((d_708_result_).error)
            return output
        elif ((d_708_result_).value).is_Full:
            output = Wrappers.Result_Success(((d_708_result_).value).data)
            return output
        elif True:
            output = Wrappers.Result_Failure(software_amazon_cryptography_materialproviders_internaldafny_types.Error_EntryDoesNotExist(_dafny.Seq("Entry does not exist")))
            return output
        return output

    def PutCacheEntry(self, input):
        output: Wrappers.Result = Wrappers.Result.default(_dafny.defaults.tuple())()
        (self.inFlight).Remove((input).identifier)
        out145_: Wrappers.Result
        out145_ = (self.wrapped).PutCacheEntry_k(input)
        output = out145_
        return output

    def DeleteCacheEntry(self, input):
        output: Wrappers.Result = Wrappers.Result.default(_dafny.defaults.tuple())()
        (self.inFlight).Remove((input).identifier)
        out146_: Wrappers.Result
        out146_ = (self.wrapped).DeleteCacheEntry_k(input)
        output = out146_
        return output

    def UpdateUsageMetadata(self, input):
        output: Wrappers.Result = Wrappers.Result.default(_dafny.defaults.tuple())()
        out147_: Wrappers.Result
        out147_ = (self.wrapped).UpdateUsageMetadata_k(input)
        output = out147_
        return output

File: AwsCryptographicMaterialProviders/runtimes/java/src/main/dafny-generated/StormTracker_Compile/StormTracker.java
// Class StormTracker
// Dafny class StormTracker compiled into Java
package StormTracker_Compile;

@SuppressWarnings({"unchecked", "deprecation"})
public class StormTracker {
  public StormTracker() {
    this.wrapped = null;
    this.inFlight = null;
    this.gracePeriod = 0L;
    this.graceInterval = 0L;
    this.fanOut = 0L;
    this.inFlightTTL = 0L;
    this.lastPrune = 0L;
    this.sleepMilli = 0L;
  }
  public LocalCMC_Compile.LocalCMC wrapped;
  public DafnyLibraries.MutableMap<dafny.DafnySequence<? extends java.lang.Byte>, java.lang.Long> inFlight;
  public long gracePeriod;
  public long graceInterval;
  public long fanOut;
  public long inFlightTTL;
  public long lastPrune;
  public long sleepMilli;
  public void __ctor(software.amazon.cryptography.materialproviders.internaldafny.types.StormTrackingCache cache)
  {
    long _0_gracePeriod = 0L;
    long _1_graceInterval = 0L;
    long _2_inFlightTTL = 0L;
    if ((((cache).dtor_timeUnits()).UnwrapOr(software.amazon.cryptography.materialproviders.internaldafny.types.TimeUnits._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.TimeUnits.create_Seconds())).is_Seconds()) {
      _0_gracePeriod = (long) (long) ((((long) ((cache).dtor_gracePeriod()))) * ((long) 1000L));
      _1_graceInterval = (long) (long) ((((long) ((cache).dtor_graceInterval()))) * ((long) 1000L));
      _2_inFlightTTL = (long) (long) ((((long) ((cache).dtor_inFlightTTL()))) * ((long) 1000L));
    } else {
      _0_gracePeriod = ((long) ((cache).dtor_gracePeriod()));
      _1_graceInterval = ((long) ((cache).dtor_graceInterval()));
      _2_inFlightTTL = ((long) ((cache).dtor_inFlightTTL()));
    }
    LocalCMC_Compile.LocalCMC _nw0 = new LocalCMC_Compile.LocalCMC();
    _nw0.__ctor(((long) ((cache).dtor_entryCapacity())), ((long) (((cache).dtor_entryPruningTailSize()).UnwrapOr(software.amazon.cryptography.materialproviders.internaldafny.types.CountingNumber._typeDescriptor(), 1))));
    (this).wrapped = _nw0;
    DafnyLibraries.MutableMap<dafny.DafnySequence<? extends java.lang.Byte>, java.lang.Long> _nw1 = new DafnyLibraries.MutableMap<dafny.DafnySequence<? extends java.lang.Byte>, java.lang.Long>(dafny.DafnySequence.<java.lang.Byte>_typeDescriptor(BoundedInts_Compile.uint8._typeDescriptor()), software.amazon.cryptography.materialproviders.internaldafny.types.PositiveLong._typeDescriptor());
    (this).inFlight = _nw1;
    (this).gracePeriod = _0_gracePeriod;
    (this).graceInterval = _1_graceInterval;
    (this).fanOut = ((long) ((cache).dtor_fanOut()));
    (this).inFlightTTL = _2_inFlightTTL;
    (this).sleepMilli = ((long) ((cache).dtor_sleepMilli()));
    (this).lastPrune = (long) 0L;
  }
  public boolean FanOutReached(long now)
  {
    boolean res = false;
    (this).PruneInFlight(now);
    res = (this.fanOut) <= (((this.inFlight).Size()).longValue());
    return res;
  }
  public long AddLong(long x, long y)
  {
    if ((x) < ((long) (long) (((StandardLibrary_mUInt_Compile.__default.INT64__MAX__LIMIT()).longValue()) - (y)))) {
      return (long) (long) ((x) + (y));
    } else {
      return (StandardLibrary_mUInt_Compile.__default.INT64__MAX__LIMIT()).longValue();
    }
  }
  public boolean GracePeriod_q(software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput result, long now)
  {
    return (((result).dtor_expiryTime()) < ((long) 9223372036854775L)) && (((long) (long) (((long) (long) (((result).dtor_expiryTime()) * ((long) 1000L))) - (this.gracePeriod))) <= (now));
  }
  public CacheState CheckInFlight(dafny.DafnySequence<? extends java.lang.Byte> identifier, software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput result, long now)
  {
    CacheState output = CacheState.Default();
    if(true) {
      boolean _0_fanOutReached;
      boolean _out0;
      _out0 = (this).FanOutReached(now);
      _0_fanOutReached = _out0;
      if (_0_fanOutReached) {
        output = CacheState.create_Full(result);
        return output;
      } else if (!((this).GracePeriod_q(result, now))) {
        output = CacheState.create_Full(result);
        return output;
      } else {
        if ((this.inFlight).HasKey(identifier)) {
          long _1_entry;
          _1_entry = (this.inFlight).Select(identifier);
          if (((this).AddLong(_1_entry, this.graceInterval)) > (now)) {
            output = CacheState.create_Full(result);
            return output;
          }
        }
        (this.inFlight).Put(identifier, now);
        output = CacheState.create_EmptyFetch();
        return output;
      }
    }
    return output;
  }
  public void PruneInFlight(long now)
  {
    if ((((this.inFlight).Size()).longValue()) < (this.fanOut)) {
      return;
    }
    if (((long) (long) ((now) - ((long) 1000L))) < (this.lastPrune)) {
      return;
    }
    (this).lastPrune = now;
    dafny.DafnySet<? extends dafny.DafnySequence<? extends java.lang.Byte>> _0_keySet;
    _0_keySet = (this.inFlight).Keys();
    dafny.DafnySequence<? extends dafny.DafnySequence<? extends java.lang.Byte>> _1_keys;
    dafny.DafnySequence<? extends dafny.DafnySequence<? extends java.lang.Byte>> _out0;
    _out0 = SortedSets.__default.<dafny.DafnySequence<? extends java.lang.Byte>>SetToSequence(dafny.DafnySequence.<java.lang.Byte>_typeDescriptor(BoundedInts_Compile.uint8._typeDescriptor()), _0_keySet);
    _1_keys = _out0;
    long _hi0 = (long) (_1_keys).cardinalityInt();
    for (long _2_i = (long) 0L; java.lang.Long.compareUnsigned(_2_i, _hi0) < 0; _2_i++) {
      long _3_v;
      _3_v = (this.inFlight).Select(((dafny.DafnySequence<? extends java.lang.Byte>)(java.lang.Object)((_1_keys).select(dafny.Helpers.unsignedToInt(_2_i)))));
      if ((now) >= ((this).AddLong(_3_v, this.inFlightTTL))) {
        (this.inFlight).Remove(((dafny.DafnySequence<? extends java.lang.Byte>)(java.lang.Object)((_1_keys).select(dafny.Helpers.unsignedToInt(_2_i)))));
      }
    }
  }
  public CacheState CheckNewEntry(dafny.DafnySequence<? extends java.lang.Byte> identifier, long now)
  {
    CacheState output = CacheState.Default();
    boolean _0_fanOutReached;
    boolean _out0;
    _out0 = (this).FanOutReached(now);
    _0_fanOutReached = _out0;
    if (_0_fanOutReached) {
      output = CacheState.create_EmptyWait();
      return output;
    } else if ((this.inFlight).HasKey(identifier)) {
      long _1_entry;
      _1_entry = (this.inFlight).Select(identifier);
      if ((now) < ((this).AddLong(_1_entry, this.graceInterval))) {
        output = CacheState.create_EmptyWait();
        return output;
      }
    }
    (this.inFlight).Put(identifier, now);
    output = CacheState.create_EmptyFetch();
    return output;
  }
  public Wrappers_Compile.Result<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error> GetFromCacheWithTime(software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryInput input, long now)
  {
    Wrappers_Compile.Result<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error> output = Wrappers_Compile.Result.<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error>Default(CacheState._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), CacheState.Default());
    if(true) {
      Wrappers_Compile.Result<software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types.Error> _0_result;
      Wrappers_Compile.Result<software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types.Error> _out0;
      _out0 = (this.wrapped).GetCacheEntryWithTime(input, dafny.DafnyEuclidean.EuclideanDivision(now, (long) 1000L));
      _0_result = _out0;
      if ((_0_result).is_Success()) {
        CacheState _1_newResult;
        CacheState _out1;
        _out1 = (this).CheckInFlight((input).dtor_identifier(), (_0_result).dtor_value(), now);
        _1_newResult = _out1;
        output = Wrappers_Compile.Result.<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error>create_Success(CacheState._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), _1_newResult);
        return output;
      } else if (((_0_result).dtor_error()).is_EntryDoesNotExist()) {
        CacheState _2_newResult;
        CacheState _out2;
        _out2 = (this).CheckNewEntry((input).dtor_identifier(), now);
        _2_newResult = _out2;
        output = Wrappers_Compile.Result.<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error>create_Success(CacheState._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), _2_newResult);
        return output;
      } else {
        output = Wrappers_Compile.Result.<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error>create_Failure(CacheState._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), (_0_result).dtor_error());
        return output;
      }
    }
    return output;
  }
  public Wrappers_Compile.Result<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error> GetFromCache(software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryInput input)
  {
    Wrappers_Compile.Result<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error> output = Wrappers_Compile.Result.<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error>Default(CacheState._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), CacheState.Default());
    if(true) {
      long _0_now;
      long _out0;
      _out0 = Time.__default.CurrentRelativeTimeMilli();
      _0_now = _out0;
      Wrappers_Compile.Result<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error> _out1;
      _out1 = (this).GetFromCacheWithTime(input, _0_now);
      output = _out1;
    }
    return output;
  }
  public Wrappers_Compile.Result<software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types.Error> GetCacheEntry(software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryInput input)
  {
    Wrappers_Compile.Result<software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types.Error> output = (Wrappers_Compile.Result<software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types.Error>)null;
    if(true) {
      Wrappers_Compile.Result<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error> _0_result;
      Wrappers_Compile.Result<CacheState, software.amazon.cryptography.materialproviders.internaldafny.types.Error> _out0;
      _out0 = (this).GetFromCache(input);
      _0_result = _out0;
      if ((_0_result).is_Failure()) {
        output = Wrappers_Compile.Result.<software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types.Error>create_Failure(software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), (_0_result).dtor_error());
        return output;
      } else if (((_0_result).dtor_value()).is_Full()) {
        output = Wrappers_Compile.Result.<software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types.Error>create_Success(software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), ((_0_result).dtor_value()).dtor_data());
        return output;
      } else {
        output = Wrappers_Compile.Result.<software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types.Error>create_Failure(software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error.create_EntryDoesNotExist(dafny.DafnySequence.asString("Entry does not exist")));
        return output;
      }
    }
    return output;
  }
  public Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> PutCacheEntry(software.amazon.cryptography.materialproviders.internaldafny.types.PutCacheEntryInput input)
  {
    Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> output = Wrappers_Compile.Result.<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error>Default(dafny.Tuple0._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), dafny.Tuple0.Default());
    if(true) {
      (this.inFlight).Remove((input).dtor_identifier());
      Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> _out0;
      _out0 = (this.wrapped).PutCacheEntry_k(input);
      output = _out0;
    }
    return output;
  }
  public Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> DeleteCacheEntry(software.amazon.cryptography.materialproviders.internaldafny.types.DeleteCacheEntryInput input)
  {
    Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> output = Wrappers_Compile.Result.<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error>Default(dafny.Tuple0._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), dafny.Tuple0.Default());
    if(true) {
      (this.inFlight).Remove((input).dtor_identifier());
      Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> _out0;
      _out0 = (this.wrapped).DeleteCacheEntry_k(input);
      output = _out0;
    }
    return output;
  }
  public Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> UpdateUsageMetadata(software.amazon.cryptography.materialproviders.internaldafny.types.UpdateUsageMetadataInput input)
  {
    Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> output = Wrappers_Compile.Result.<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error>Default(dafny.Tuple0._typeDescriptor(), software.amazon.cryptography.materialproviders.internaldafny.types.Error._typeDescriptor(), dafny.Tuple0.Default());
    if(true) {
      Wrappers_Compile.Result<dafny.Tuple0, software.amazon.cryptography.materialproviders.internaldafny.types.Error> _out0;
      _out0 = (this.wrapped).UpdateUsageMetadata_k(input);
      output = _out0;
    }
    return output;
  }
  private static final dafny.TypeDescriptor<StormTracker> _TYPE = dafny.TypeDescriptor.<StormTracker>referenceWithInitializer(StormTracker.class, () -> (StormTracker) null);
  public static dafny.TypeDescriptor<StormTracker> _typeDescriptor() {
    return (dafny.TypeDescriptor<StormTracker>) (dafny.TypeDescriptor<?>) _TYPE;
  }
  @Override
  public java.lang.String toString() {
    return "StormTracker.StormTracker";
  }
}
